File: expm/python/uProject.py
#import
import matplotlib.pyplot as plt
import math as m

#-------------------------------------------------------------------------------

#parameters
nbcol_lines = 2
step = 1
begin = 0
end = 2
color = "red"

a = [[-12, 12], [3, -3]]

#-------------------------------------------------------------------------------

#expm function
def initmat(X):
    """Init the matric with 0"""
    l = len(X)
    result = []
    for i in range(l):
        result.append([0] * l)
    return result

def power2(X):
    """Power of two square matrix"""
    l = len(X)
    result = initmat(X)
            
    for i in range(l):
        for j in range(l):
            for k in range(l):
                result[i][j] += X[i][k] * X[k][j]
    return result

def swapmat(X, Y):
    """Change two matrix"""
    l = len(X)
    for i in range(l):
        for j in range(l):
            X[i][j] = Y[i][j]
            Y[i][j] = 0
    return (X, Y)
            
def powermat(X, p):
    """Power of square matrix"""
    l = len(X)
    m = 2
    result = initmat(X)
    
    Y = power2(X)
    
    while m < p:
        for i in range(l):
            for j in range(l):
                for k in range(l):
                    result[i][j] += Y[i][k] * X[k][j]
                
        Y, result = swapmat(Y, result)
        m = m + 1
        
    return Y

def facto(n):
    """Computes n factorial, begins 3"""
    result = 1
    for i in range(2, n + 1):
        result = result * i
    return result

def matxl(X, n):
    """Computes the multiplication between a square matrix and a number"""
    l = len(X)
    result = initmat(X)
    
    for i in range(l):
        for j in range(l):
            result[i][j] = X[i][j] * n
    return result

def addmat(X, Y):
    """Computes the additional of two matrix"""
    l = len(X)
    result = initmat(X)
    
    for i in range(l):
        for j in range(l): 
            result[i][j] = X[i][j] + Y[i][j]
    return result

def matid(X):
    """Returns the identity matrix of the X matrix"""
    l = len(X)
    result = initmat(X)
        
    for i in range(l):
        for j in range(l):
            if(i == j):
                result[i][j] = 1
                
    return result
    
def expmat(X, value):
    """Computes exponential of matrix (O until value)"""
    l = len(X)
    result = initmat(X)
        
    for k in range(2, value):
        result = addmat(matxl(powermat(X, k),(1 / facto(k))), result)
        
    result = addmat(result, X)
    result = addmat(result, matid(X))
    
    return result

def probability(X, t, n):
    """Final function"""
    Y = matxl(X, t)
    Y = expmat(Y, n)
    return Y

#-------------------------------------------------------------------------------
    
#graphics settings
plt.xlim(0, 3)
plt.ylim(0, 1)

plt.grid(b=None, axis='both', linestyle = '--')
plt.title("Micro project 1")
plt.xlabel("Time")
plt.ylabel("Probability")

fig = plt.gcf()
fig.set_size_inches(17.5, 8.5)

#-------------------------------------------------------------------------------

#points
vect = []

for i in range(nbcol_lines * nbcol_lines):
    vect.append([])

for i in range(begin, end * 100, step):
    m = probability(a, i / 100, 100)
    cpt = 0
    
    for j in range(nbcol_lines):
        for k in range(nbcol_lines):
            vect[cpt].append(m[j][k])
            cpt = cpt + 1

x = [i / 100 for i in range(begin, end * 100, step)]

#-------------------------------------------------------------------------------

#display
for i in range(nbcol_lines * nbcol_lines):
    plt.plot(x, vect[i], color="red")

plt.savefig('graph.pdf')
plt.savefig('graph.png')

File: expm/r/uProject.R
#import
library(Matrix)

#-------------------------------------------------------------------------------

#parameters
nbccol_lines <- 2
step <- 0.01
begin <- 0
end <- 200
color <- "red"

a <- matrix(c(-12, 12, 3, -3), nbccol_lines, nbccol_lines)

#-------------------------------------------------------------------------------

#points
vect <- matrix(list(), nrow = 1, ncol = (nbccol_lines*nbccol_lines))

for (i in seq(begin, end, by = step))
{
    m <- Matrix::expm(a*i)
    cpt <- 1
    
    for (j in 1:nbccol_lines)
    {
        for (k in 1:nbccol_lines)
        {
            vect[[1, cpt]] <- c(vect[[1, cpt]], m[j, k])
            cpt <- cpt + 1
        }
    }
}

x <- seq(begin, end, by = step)

#-------------------------------------------------------------------------------

#display
jpeg("graph.jpg")
plot(x, vect[[1, 1]], main = "Micro Project 1", "l", col = color, xlab = "Time",
    ylab = "Probability", xlim = c(begin, end*step), ylim = c(0,1))

for (i in 2:(nbccol_lines*nbccol_lines))
{
    lines(x, vect[[1, i]], col = color)
}

abline(v = 0:10,  lty = 2, col = "grey")
abline(h = 0:10,  lty = 2, col = "grey")
